test/e2e/framework/service_util.go

```diff
@@ -806,6 +807,108 @@ func (j *ServiceTestJig) waitForPodsReady(namespace string, pods []string) error
+
+		// If .spec.clusterIP is set to "" or "None", ClusterIP is not created, so can not test reachability over clusterIP:servicePort
+		isClusterIPV46, err := regexp.MatchString(RegexIPv4+"||"+RegexIPv6, clusterIP)
+		if err != nil {
+			e2elog.Logf("Unable to parse ClusterIP: %s", clusterIP)
+			return err
+		}
+		ep, err := j.Client.CoreV1().Endpoints(f.Namespace.Name).Get(svc.Name, metav1.GetOptions{})
+		if err != nil {
+			e2elog.Logf("Unable to get endpoints for service: %s", svc.Name)
+			return err
+		}
+		// Endpoints must be available for ClusterIP service. (reachable service)
+		if len(ep.Subsets) == 0 {
+			e2elog.Logf("Service endpoints are not available for service: %s", svc.Name)
+			return nil
+		}
+		// service should be reachable on servicename and clusterip over all defined service ports
+		// Endpoints using servicePort
+		for _, p := range ports {
+			svcEndpoints = append(svcEndpoints, net.JoinHostPort(svc.Name, strconv.Itoa(int(p.Port))))
+			if isClusterIPV46 {
+				svcEndpoints = append(svcEndpoints, net.JoinHostPort(clusterIP, strconv.Itoa(int(p.Port))))
+			}
+		}
+		for _, e := range svcEndpoints {
+			cmds = append(cmds, fmt.Sprintf("wget -qO- -T 2 http://%s", e))
+		}
+
+		for _, cmd := range cmds {
+			_, err := RunHostCmd(pod.Namespace, pod.Name, cmd)
+			ExpectNoError(err, "Service is not reachable on following endpoint")
+		}
+	case v1.ServiceTypeNodePort:
+		// Check if serviceEndpoints exists
+		ep, err := j.Client.CoreV1().Endpoints(f.Namespace.Name).Get(svc.Name, metav1.GetOptions{})
+		if len(ep.Subsets) == 0 {
+			e2elog.Logf("Service endpoints not available for service: %s", svc.Name)
+			return nil
+		}
+		// Endpoints must be available for nodePort service.
+		if err != nil {
+			e2elog.Logf("Unable to get endpoints for service: %s", svc.Name)
+			return err
+		}
+		nodes := GetReadySchedulableNodesOrDie(j.Client)
+		internalAddrs := e2enode.CollectAddresses(nodes, v1.NodeInternalIP)
+		externalAddrs := e2enode.CollectAddresses(nodes, v1.NodeExternalIP)
+		// Endpoints using nodePort
+		for _, p := range ports {
+			svcEndpoints = append(svcEndpoints, net.JoinHostPort(svc.Name, strconv.Itoa(int(p.Port))))
+			svcEndpoints = append(svcEndpoints, net.JoinHostPort(clusterIP, strconv.Itoa(int(p.Port))))
+
+			for _, internalAddr := range internalAddrs {
+				svcEndpoints = append(svcEndpoints, net.JoinHostPort(internalAddr, strconv.Itoa(int(p.NodePort))))
+			}
+			for _, externalAddr := range externalAddrs {
+				svcEndpoints = append(svcEndpoints, net.JoinHostPort(externalAddr, strconv.Itoa(int(p.NodePort))))
+			}
+		}
+		for _, e := range svcEndpoints {
+			cmds = append(cmds, fmt.Sprintf("wget -qO- -T 2 http://%s", e))
+		}
+
+		for _, cmd := range cmds {
+			_, err := RunHostCmd(pod.Namespace, pod.Name, cmd)
+			ExpectNoError(err, "Service is not reachable on following endpoint")
+		}
+
+	case v1.ServiceTypeExternalName:
+		// Service should resolve to IP
+		cmd := fmt.Sprintf("nslookup %s", svc.Name)
```

Comment:
>suggest `dig` or `host` instead.

Seems in agnhost:2.2, `dig` and `host` are not available. 
FYI, List of all available packages are - 
```
alpine-baselayout-3.0.4-r0 - Alpine base dir structure and init scripts
alpine-keys-2.1-r1 - Public keys for Alpine Linux packages
apk-tools-2.7.6-r0 - Alpine Package Keeper - package manager for alpine
bash-4.3.48-r1 - The GNU Bourne Again shell
busybox-1.26.2-r11 - Size optimized toolbox of many common UNIX utilities
ca-certificates-20161130-r3 - Common CA certificates PEM files
curl-7.61.1-r2 - URL retrival utility and library
iperf-2.0.9-r1 - A tool to measure IP bandwidth using UDP or TCP
iproute2-4.10.0-r1 - IP Routing Utilities
iptables-1.6.1-r0 - Linux kernel firewall, NAT and packet mangling tools
libbsd-0.8.3-r3 - commonly-used BSD functions not implemented by all libcs
libc-utils-0.7.1-r0 - Meta package to pull in correct libc
libcurl-7.61.1-r2 - The multiprotocol file transfer library
libelf-0.8.13-r2 - libelf is a free ELF object file access library
libgcc-6.3.0-r4 - GNU C compiler runtime libraries
libmnl-1.0.4-r0 - Library for minimalistic netlink
libnftnl-libs-1.0.7-r0 - Netfilter library providing interface to the nf_tables subsystem (libraries)
libressl2.5-libcrypto-2.5.5-r2 - libressl libcrypto library
libressl2.5-libssl-2.5.5-r2 - libressl libssl library
libressl2.5-libtls-2.5.5-r2 - libressl libtls library
libssh2-1.8.2-r0 - library for accessing ssh1/ssh2 protocol servers
libstdc++-6.3.0-r4 - GNU C++ standard runtime library
musl-1.1.16-r15 - the musl c library (libc) implementation
musl-utils-1.1.16-r15 - the musl c library (libc) implementation
ncurses-libs-6.0_p20171125-r1 - Ncurses libraries
ncurses-terminfo-6.0_p20171125-r1 - Console display library
ncurses-terminfo-base-6.0_p20171125-r1 - Descriptions of common terminals
netcat-openbsd-1.130-r1 - The TCP/IP swiss army knife. OpenBSD variant.
readline-6.3.008-r5 - GNU readline library
scanelf-1.2.2-r0 - Scan ELF binaries for stuff
ssl_client-1.26.2-r11 - EXternal ssl_client for busybox wget
zlib-1.2.11-r0 - A compression/decompression Library
```

>What if the external name the service in question points to is bogus?

We are expecting that externalName added to service must not be bogus and check must fail in case of bogus (Hope, Externalname resolving to atleast 1 IP is sufficient to test reachability/nameResolution)

Reply:
Any suggestion on replacement of `nslookup` (or use other binaries from agnhost). Once `dig` or `host` are added to agnhost, we can replace.

Reply:
Well, at least do -q=A which tells it to only lookup the A record not the AAAA record. I have seen instances where it will lookup both and then if the AAAA returns empty before the A it gets confused.

Reply:
agnhost-pause:2.2 is not accepting any query types `-q=A` for `nslookup` command. It's failing -
https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/79667/pull-kubernetes-e2e-gce/1149783426329481217/

Reply:
I suspect agnhost-pause:2.2 accepts only `nslookup [HOST] [SERVER]` cmd format.

Reply:
Ugh. Ok. For Linux, the APK `bind-tools` contains these. For Windows, it looks like you can get them from https://www.isc.org/download/

Can you add an issue to update agnhost to install `dig` and to update these to use `dig`? Let's not continue to hold these up on that.

Reply:
Tracking in #80194.

Reply:
Just found, `bind-tools-9.11.6_p1-r1` is already included in `agnhost:2.4`. Let me update PR to use `dig`.

Reply:
I see, `imageutils.Agnhost` resolves to `gcr.io/kubernetes-e2e-test-images/agnhost:2.4` which has `dig 9.11.6-P1`. 
Surprisingly following error is encountered [here](https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/79667/pull-kubernetes-e2e-gce/1150979070255697921/) (in `kind` cluster too)
``` dig: not found\ncommand terminated with exit code 127\n\nerror:\nexit status 127```


Following is the output of anghost pod from local `kind` cluster.
```
➜  kubernetes (master-service) ✔ k get pod/agnhost --template={{.status.containerStatuses}}
[map[containerID:containerd://a076f2331c49ab35585a723a06169e00fc70f67a6a3b0ed07f8c2ff33bf3fcaa image:gcr.io/kubernetes-e2e-test-images/agnhost:2.4 imageID:gcr.io/kubernetes-e2e-test-images/agnhost@sha256:080e89220a20400cefd1876ac2fddc0f379dccda60cd7d8d8d3415e1a959cd1b lastState:map[] name:agnhost ready:true restartCount:0 state:map[running:map[startedAt:2019-07-16T02:29:28Z]]]]% 
➜  kubernetes (master-service) ✔ k exec agnhost dig version

; <<>> DiG 9.11.6-P1 <<>> version
```

Reply:
https://github.com/kubernetes/kubernetes/blob/master/test/images/agnhost/Dockerfile does not include `bind-tools`, so somehow your local image is different. Let's just stick with `nslookup` for now until that gets worked out.

Reply:
Yes, seems your local image is different:

```
jbelamaric@jbelamaric:~$ k run --restart=Never --image gcr.io/kubernetes-e2e-test-images/agnhost:2.4 agnhost pause
pod/agnhost created
jbelamaric@jbelamaric:~$ k exec -it agnhost dig
rpc error: code = 2 desc = oci runtime error: exec failed: container_linux.go:247: starting container process caused "exec: \"dig\": executable file not found in $PATH"

command terminated with exit code 126
jbelamaric@jbelamaric:~$ k exec -it agnhost nslookup
BusyBox v1.26.2 (2018-05-30 13:51:20 GMT) multi-call binary.

Usage: nslookup [HOST] [SERVER]

Query the nameserver for the IP address of the given HOST
optionally using a specified DNS server
command terminated with exit code 1
```

Reply:
Thanks for pointing. Having `agnhost version` support in image would be really helpful.

Reply:
Reverted back to `nslookup`.